streams/shared/src/main/scala/zio/stream/ZTransducer.scala

```diff
@@ -192,6 +192,61 @@ object ZTransducer {
       case None     => ZIO.succeedNow(Chunk.empty)
     }
 
+  /**
+   * Creates a transducer that chunks individual chunks.
+   */
+  def chunkChunks[A]: ZTransducer[Any, Nothing, A, Chunk[A]] =
+    identity[A].mapChunks(chunk => if (chunk.isEmpty) Chunk.empty else Chunk.single(chunk))
+
+  /**
+   * Creates a transducer that limits chunks to the given size.
+   */
+  def chunkLimit[A](max: Int): ZTransducer[Any, Nothing, Chunk[A], Chunk[A]] = {
+
+    @scala.annotation.tailrec
+    def make(in: Chunk[A], out: Chunk[Chunk[A]]): Chunk[Chunk[A]] =
+      if (in.length <= max) out + in else make(in.drop(max), out + in.take(max))
+
+    @scala.annotation.tailrec
+    def go(in: Chunk[Chunk[A]], out: Chunk[Chunk[A]]): Chunk[Chunk[A]] =
+      (in.length: @scala.annotation.switch) match {
+        case 0                          => out
+        case 1 if in.head.length <= max => out ++ in
+        case _                          => go(in.tail, make(in.head, out))
+      }
+
+    ZTransducer.identity[Chunk[A]].mapChunks(go(_, Chunk.empty))
+  }
+
+  /**
+   * Creates a transducer that emits chunks of a fixed `size`.
+   * The `pad` function is called on the last chunk if it does not have sufficient elements.
+   */
+  def chunkN[A](size: Int, pad: Chunk[A] => Chunk[A]): ZTransducer[Any, Nothing, Chunk[A], Chunk[A]] =
+    apply(ZRefM.make[Chunk[A]](Chunk.empty).toManaged_.map { ref =>
+      @scala.annotation.tailrec
+      def go(as: Chunk[A], in: Chunk[Chunk[A]], out: Chunk[Chunk[A]]): (Chunk[Chunk[A]], Chunk[A]) =
+        if (in.isEmpty)
+          if (as.length < size) (out, as) else go(as.drop(size), in, out + as.take(size))
+        else if (as.length < size) go(as ++ in.head, in.tail, out)
```

Comment:
Incrementally building up chunks with `++` can create very inefficient chunks, because we currently don't balance the `Concat` tree. Could you convert this to a loop that builds up the chunk using a `ChunkBuilder`?

Reply:
Done